_posts/2024-05-11-coroutines-and-deadlocks.md

```diff
@@ -0,0 +1,231 @@
+    launch {
+        delay(1000L) // non-blocking delay for 1 second (simulating a long-running task)
+        println("World!")
+    }
+    // Main thread continues while the coroutine is delayed
+    println("Hello,") // This line is executed first
+}
+```
+
+Resulting output:
+```bash
+Hello,
+World!
+```
+
+Although it seems counterintuitive to use a blocking call in a discussion about non-blocking code, runBlocking serves a specific purpose. 
+It acts as a bridge allowing non-coroutine code to interact seamlessly with libraries that are designed using coroutines, making it useful primarily for main functions and tests
+
+## Executing Blocking Code with Coroutines
+
+Although coroutines are non-blocking, there are situations where you need to call blocking code, such as database operations or network requests that do not support asynchronous APIs. 
+To prevent blocking the main thread or the coroutine context threads, you can use the Dispatchers.IO context. This dispatcher is optimized for offloading blocking IO tasks to a shared pool of threads.
+
+```kotlin
+// Launch a new coroutine in the IO context for blocking operations
+withContext(Dispatchers.IO) {
+    val result = performBlockingOperation()
+    println("Result: $result")
+}
+
+// Simulates a blocking operation
+fun performBlockingOperation(): String {
+    Thread.sleep(2000) // Simulates a long-running blocking task
+    return "Operation Complete"
+}
+```
+
+If you are not familiar with Kotlin, you may ask what is the difference between `Thread.sleep(...)` and the suspending function `delay(...)`. 
+While `Thread.sleep` actually blocks the thread and makes it unresponsive (the thread is literally sleeping and doing nothing), 
+`delay` is a suspending function that suspends the coroutine and releases the executing thread, making it available to perform other work. 
+Eventually, the coroutine is rescheduled, and a thread is assigned to resume its execution. 
+With coroutines, we do not know which thread will execute it when it resumes. 
+This makes coroutines much more efficient and scalable for handling asynchronous tasks.
+
+# Exposed and Coroutine
+
+Exposed provides a mechanism intended to bridge its inherently blocking JDBC operations with Kotlin's non-blocking coroutines, 
+aiming to streamline integration in coroutine-based environments. 
+While Exposed uses JDBC under the hood, which is blocking by design, it offers utilities to adapt to Kotlin's asynchronous programming model.
+
+Here’s how you can transition from a traditional transaction to one that is better integrated with coroutines:
+
+```kotln
+// Traditional blocking transaction
+val count = transaction {
+    FooTable.selectAll().count()
+}
+
+// Coroutine-friendly
+val count = newSuspendedTransaction(Dispatchers.IO) {
+    FooTable.selectAll().count()
+}
+```
+
+While this feature facilitates the use of coroutines by superficially adapting to their syntax and style, 
```

Comment:
The first one is normal code without coroutines. There are no coroutines returned or created here. It just about the context -> who is responsible to execute the code block.
Since FooTable.selectAll().count() is a blocking call (IO), without coroutines the executing thead is blocked, until the operation completes. 
Within 
```
 newSuspendedTransaction(Dispatchers.IO) {
    FooTable.selectAll().count()
}
```
the thread is also blocked, but it a special IO Thread and not the initating thread, for example the UI Main thread, which should never been blocked.

Reply:
now i understand. this should probably go into the blog post? :)